### PR TITLE
Add spin test for collision

### PR DESCRIPTION
**Pull request by shivaang12**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   |  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on |  |

---

## Description of contribution in a few bullet points
* Add collision case for spin recovery

## Description of documentation updates required from your changes

---

## Future work that may be required in bullet points
* Change expecting status to FAILED (or false in bool) after action status fix in #1779

### CODE REVIEW

File: nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp [hunk 1]
```diff
@@ -64,6 +65,8 @@ class SpinRecoveryTester
 private:
   void sendInitialPose();
 
+  void sendObstaclePose();
```
Comment (SteveMacenski) [inline nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: I think this should be renamed `sendPoseUnderObstacle()`. I was a little confused thinking that you were actually publishing to a topic to move a gazebo object under the robot, which would have also worked.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: I just found out that changing the pose by the topic will not work as the actual gazebo server is running. I think I have to create a new map where robot is spawn on obstacle or I have make robot go near one after non-collision spin tests. What do you think would be best?

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: I don't know what you mean. You can move the pose however you like via the AMCL initial pose topic.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: I am trying to say that it isn't working as we are also running headless gazebo in the background. No matter what pose I published, it gets corrected to the spawned coordinates. Even if we change the pose with AMCL, its not reflecting because the gazebo is running the background. To fix it I tried changing the pose in the world file and it works. In conclusion, I want to know which would seem more appropriate, making separate world file or somehow make robot travel to the point where I want after normal spin test.

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: That makes no sense. Please look into it. Gazebo has nothing to do with this, nor does headless gazebo play a role.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: Because Spin recovery relies on "odom" topic for position, no matter how we change AMCL pose it won't affect the collision checking because it uses coordinates from "odom" topic to check for collision. And I think "odom" topic is publishing robots position through gazebo plugin. This why I said its gazebo.

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: When you move it though you should be physically moving the robot in the map frame so any sensor data, inflation, and static map will have collisions found.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: Yes, you move the robot in the map but that won't update the values in the "odom" topic. "odom" is independent it only updates on the input to "cmd_vel". So when spin recovery checks for collision, first it gets the current coordinates of the robot from "odom" topic and it simulates (or looks further in future) for collision. This is why updating AMCL pose won't do it. Please refer to these links for code lines.

Spin recovers collision checker https://github.com/ros-planning/navigation2/blob/13d6629356e897e1717f2fd9579d807bd3c8cfc6/nav2_recoveries/plugins/spin.cpp#L144-L168

Below code snippets shows global frame (which is odom instead of map) and base frame
https://github.com/ros-planning/navigation2/blob/13d6629356e897e1717f2fd9579d807bd3c8cfc6/nav2_recoveries/src/recovery_server.cpp#L49-L54

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: I dont understand the problem, if you physically transport your robot to a pose on top of a static obstacle, the static layer will result it being in collision with something.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: Yes, but changing AMCL pose will not result in physical transported. "Odom" is independent from AMCL.

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: Pleaes use the initial pose tool in rviz, you will see that it works. If you don't feel this will work, then please close the PR.

Comment (shivaang12) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: @SteveMacenski, I am sorry for not being clear in my defense. I will share a screenshot of my logs. Also, if you claim it should work that means what I have done in this PR should also work (i.e. changing the AMCL pose) and the obstacle detection line in `spin.cpp` should get covered right in codecov? But it does not get covered despite the provided pose is in direct collision. I will share a screenshot of my logs.


The first line in the log confirms the setting of position. (current pose is X=-2.0 and Y=-0.5 and new pose is X=0.0 and Y=0.0)
The second and third line confirms from AMCL that pose has been received and spits out which pose has been set (now current pose is X=0.0 Y=0.0)
But the last line of the log still says pose is X=-2.0 Y=-0.5 despite being change earlier to 0.0

Hope this explanation sheds some light to the issue.

Also, the initial pose tool in rviz sets tf between "base_frame" and "map" not odom. Nav2_recovery global frame is "odom" while Nav2_amcl global_frame is "map". So when you change the AMCL pose, the frame is "map" not "odom". You can refer to nav2_params.yaml file to confirm.

In short, the whole nav2_recovery operates on odom frame. And changing pose in "map" (AMCL pose) won't change anything. I have tested the whole day on this and this is my conclusion.

Also, idu how laser scan will detect an obstacle while performing spin recovery?

Also, I wanted to clear the motive behind this test, is it to cover these below lines?
https://github.com/ros-planning/navigation2/blob/13d6629356e897e1717f2fd9579d807bd3c8cfc6/nav2_recoveries/plugins/spin.cpp#L133-L137

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: > The first line in the log confirms the setting of position. (current pose is X=-2.0 and Y=-0.5 and new pose is X=0.0 and Y=0.0) The second and third line confirms from AMCL that pose has been received and spits out which pose has been set (now current pose is X=0.0 Y=0.0) But the last line of the log still says pose is X=-2.0 Y=-0.5 despite being change earlier to 0.0

They are in different reference frames. The spin is looking at odometry frame and the AMCL is on map frame, so they would be different and the odom position wouldn't change. What would change though it that since the robot is transported there, it should see static map obstacle next to / in collision with that should be inflated causing a failure. The odometry value shouldn't matter since its been virtually transported.

Comment (SteveMacenski) [reply nav2_system_tests/src/recoveries/spin/spin_recovery_tester.hpp:68]: If you don't think that'll work - we should close the PR and revisit this at some other point, it might be a little moot anyhow since we can't get a "FAILED" back on a failed recovery. Maybe the local costmap in that test doesn't use the static layer or somethign?

**Comment by shivaang12** [comment]
Agreed! I think I'll come back when the result thing is fixed.
